ipn/ipnlocal/local.go

```diff
@@ -6449,13 +6449,35 @@ func (n lastSuggestedExitNode) asAPIType() (res apitype.ExitNodeSuggestionRespon
 	return res, ErrUnableToSuggestLastExitNode
 }
 
+// processAllowedCandidates converts a string slice of exit node candidate IDs to a map of strings and ints. Int value represents nothing and won't be used.
+// Used for a faster lookup comparison in suggestExitNode.
+func processAllowedCandidates(allowedCandidates []string) (candidates set.Set[string]) {
+	for _, candidate := range allowedCandidates {
+		candidates.Add(candidate)
+	}
+	return candidates
+}
+
 func suggestExitNode(report *netcheck.Report, netMap *netmap.NetworkMap, r *rand.Rand) (res apitype.ExitNodeSuggestionResponse, err error) {
 	if report.PreferredDERP == 0 {
 		return res, ErrNoPreferredDERP
 	}
+	var allowedCandidates set.Set[string]
+	if allowed, _ := syspolicy.GetStringArray(syspolicy.AllowedSuggestedExitNodes, []string{}); len(allowed) != 0 {
+		allowedCandidates = processAllowedCandidates(allowed)
+	}
 	candidates := make([]tailcfg.NodeView, 0, len(netMap.Peers))
+	isAuto := netMap.SelfNode.CapMap().Has(tailcfg.NodeAttrAutoExitNode)
 	for _, peer := range netMap.Peers {
-		if peer.CapMap().Has(tailcfg.NodeAttrSuggestExitNode) && tsaddr.ContainsExitRoutes(peer.AllowedIPs()) {
+		var isCandidate bool
+		if len(allowedCandidates) != 0 && isAuto {
+			if allowedCandidates.Contains(string(peer.StableID())) {
+				isCandidate = true
+			}
+		} else {
+			isCandidate = true
+		}
+		if peer.CapMap().Has(tailcfg.NodeAttrSuggestExitNode) && tsaddr.ContainsExitRoutes(peer.AllowedIPs()) && isCandidate {
```

Comment:
```suggestion
		if isAuto && allowedCandidates != nil && !allowedCandidates.Contains(string(peer.StableID())) {
			continue
		}
		if peer.CapMap().Has(tailcfg.NodeAttrSuggestExitNode) && tsaddr.ContainsExitRoutes(peer.AllowedIPs()) {
```

Note the `allowedCandidates != nil` which is subtly different from `len(allowedCandidates) != 0`: if the policy is set but has no entries, that should probably mean that nothing is eligible. If the policy is unset, then that should mean that anything is eligible.

I was tempted to make `allowedCandidates` a `set.Set[tailcfg.StableNodeID]` so the cast to string isn't required, but the policy contains unverified strings and it would mean that `set.SetOf` or `set.Of` wouldn't be usable because a type conversion would need to happen, so this way feels less icky.